### PR TITLE
Juniper Contrail SDN support for OpenShift

### PR DESCRIPTION
**Pull request by savithruml**

### CODE REVIEW

**Comment by rajatchopra** [comment]
@savithruml PR needs rebase
Also, PR should likely be made to tip of master branch first and then possibly cherry-picked for 3.7 !
We are going to roll out release 3.9 soon anyway.

File: roles/openshift_loadbalancer/templates/haproxy.cfg.j2 [hunk 1]
```diff
@@ -80,3 +79,33 @@ backend {{ backend.name }}
     server      {{ server.name }} {{ server.address }} {{ server.opts }}
 {% endfor %}
 {% endfor %}
+
+{% if openshift_use_contrail | default(false) | bool %}
```
Comment (rajatchopra) [inline roles/openshift_loadbalancer/templates/haproxy.cfg.j2]: We should not use the apiserver load balancer to serve contrail UI. This should likely be its own load balancer.

File: playbooks/adhoc/uninstall.yml [hunk 2]
```diff
@@ -539,3 +562,17 @@
     # - /etc/systemd/system/haproxy.service.d
     - /etc/systemd/system/haproxy.service.d/limits.conf
     - /etc/systemd/system/haproxy.service
+
+# Remove Contrail vRouter
+- hosts: nodes
+  become: yes
+  tasks:
+  - name: Find nodes which has vRouter
+    shell: "ip addr | grep -i vhost0 | cat"
```
Comment (rajatchopra) [inline playbooks/adhoc/uninstall.yml]: Isn't there a better way to find that out?

File: roles/contrail_node/tasks/main.yaml [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+- name: Launch the Contrail installer
+  command: oc create -f /tmp/contrail-installer.yaml
+  run_once: True
+  delegate_to: "{{ contrail_master }}"
+  ignore_errors: True
+
+- name: Patch restricted scc
+  command: >
+    oc patch scc restricted --patch='{ "runAsUser": { "type": "RunAsAny" } }'
```
Comment (rajatchopra) [inline roles/contrail_node/tasks/main.yaml]: @sdodson Review this line please. I am not sure if this is acceptable to  patch the scc globally.

File: roles/contrail_node/tasks/main.yaml [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+  - name: Concatenate SSL certs
+    assemble:
+      src: "{{ websslpath }}"
+      dest: "{{ websslpath }}/contrail-webui.pem"
+      remote_src: True
+    run_once: True
+    delegate_to: "{{ groups.lb.0 }}"
+
+  - name: Restart haproxy
```
Comment (rajatchopra) [inline roles/contrail_node/tasks/main.yaml]: We need to discuss this. Not sure if the apiserver load balancer can be used directly.

File: roles/openshift_loadbalancer/templates/haproxy.cfg.j2 (hunk 1 above)
Comment (savithruml) [reply roles/openshift_loadbalancer/templates/haproxy.cfg.j2]: Hi @rajatchopra will remove this section. If we want to provide HA services for Contrail, then do you suggest bringing in our own loadbalancer node?

File: playbooks/adhoc/uninstall.yml (hunk 2 above)
Comment (savithruml) [reply playbooks/adhoc/uninstall.yml]: Yes, will use a module.

**Comment by savithruml** [comment]
@rajatchopra yes, currently our 4.X containers are packaged in a tgz which can be downloaded from juniper.net site. Our next release will host containers on dockerhub.

Also, should I rebase & commit to master? I noticed that master branch have some changes like the playbooks have been moved to a directory called as [private](https://github.com/openshift/openshift-ansible/tree/master/playbooks/openshift-master/private) which was not present in <=3.7 & sanity also checks [this file](https://github.com/openshift/openshift-ansible/blob/master/roles/lib_utils/action_plugins/sanity_checks.py) for SDN. Do we need to align with this for 3.9?

**Pull request closed without merging**